### PR TITLE
Refactor translation unit partitioning/collection as a query

### LINKED ISSUE #44486: incr.comp.: Turn translation item collection and CGU partitioning into a query.
Right now translation item collection and CGU partitioning are run as part of `rustc_trans::base::trans_crate()`. For better integration with red/green tracking this functionality should be moved into a pair of queries:
* The first query does all of translation item collection *and* partitioning and returns `(set of all trans-items, set of all partitions)`. It can have a query key of `()` as it only takes other queries as input.
* The second query has query-key `cgu_name: &str` and returns the single `CodegenUnit` with that name. It just reads from the result of the first query. It's purpose is to provide a "firewall" so codegen units can be re-used if they don't change.

Doing this refactoring will involve moving `TransItem` and `CodegenUnit` to `librustc`, so they can be used in query results. One will also have to provide a `HashStable` implementation for the results of the two queries. The implementation for `CodegenUnit` should do something like [CodegenUnit::compute_symbol_name_hash()](https://github.com/rust-lang/rust/blob/a0b34199ba6f7c922e7832160328c2a4e90b6ac6/src/librustc_trans/partitioning.rs#L176). For translation items one could maybe just hash the symbol name. 

cc @nikomatsakis @alexcrichton

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is targeted at #44486 with the ultimate goal of making the `collect_and_partition_translation_items` function a query. This mostly just involved query-ifying a few other systems along with plumbing the tcx instead of `SharedCrateContext` in a few locations.

Currently this only tackles the first bullet of #44486 and doesn't add a dedicated query for a particular codegen unit. I wasn't quite sure how to do that yet but figured this was good to put up.

Closes #44486

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Thanks, Alex! At first I wasn't sure if I liked the `CodegenUnitExt`/`TransItemExt` construction but now I think it's a pretty good solution. 

My only nit would be that the fields of `CodegenUnit` start with an underscore now. I interpret this as "these are actually private". I'd prefer if we did not need that (which, I guess, could be done by using a local builder type in the `partitioning` module). I don't think this should block this PR though.

r=me with after conflicts have been resolved.

**Comment by alexcrichton** [comment]
Oh sure that's fine, I'll add some method accessors in rustc and use those throughout librustc_trans?

**Comment by michaelwoerister** [comment]
Yes, that should work.

**Comment by alexcrichton** [comment]
Ok I've pushed [another commit](https://github.com/rust-lang/rust/pull/44529/commits/4cb1f3ed702eef5f22c65eac2abe80110ed6ef16) which changes again how exported symbols are dealt with. Everything should now be behind queries and moving us one step closer to 'codegen as a query' now that `CrateContext` no longer has a set of all exported symbols to peek at.

**Comment by alexcrichton** [comment]
re-r? @michaelwoerister

**Comment by alexcrichton** [comment]
Alright and to round this off some more, the last commits now make codegen 100% query-ified.

File: src/librustc_trans/back/write.rs [hunk 1]
```diff
@@ -666,19 +669,40 @@ fn need_crate_bitcode_for_rlib(sess: &Session) -> bool {
     sess.opts.output_types.contains_key(&OutputType::Exe)
 }
 
-pub fn start_async_translation(sess: &Session,
+pub fn start_async_translation(tcx: TyCtxt,
```
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: Nice cleanup!

File: src/librustc_trans/callee.rs [hunk 2]
```diff
@@ -109,10 +108,43 @@ pub fn get_fn<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
...
+        //  * This is a symbol defined in a foreign crate and we're just
+        //    monomorphizing in another codegen unit. In this case this symbols
+        //    is for sure not exported, so both codegen units will be using
+        //    hidden visibility. Hence, we apply a hidden visibility here.
+        //
+        //  * This is a symbol defined in our local crate. If the symbol in the
+        //    other codegen unit is also not exported then like with the foreign
+        //    case we apply a hidden visibility. If the symbol is exported from
+        //    the foreign object file, however, then we leave this at the
+        //    default visibility as we'll just import it naturally.
```
Comment (michaelwoerister) [inline src/librustc_trans/callee.rs]: Great comment!

**Comment by michaelwoerister** [comment]
@bors r+ 

Thanks! Excellent `:)`

**Pull request merged**